File: kelly/스택 & 큐/Trapping Rain Water/solution02.py
class Solution:
    def trap(self, height: List[int]) -> int:
        answer = [0] * len(height)
        stack = []

        for i in range(len(height)):
            if not stack or height[stack[-1]] >= height[i]:
                stack.append(i)
                continue

            last_popped = None
            while stack and height[stack[-1]] < height[i]:
                last_popped = stack.pop()

            low_wall = last_popped if not stack else i
            start = last_popped if not stack else stack[-1]
            for j in range(start + 1, i):
                answer[j] = height[low_wall] - height[j]

            stack.append(i)

        return sum(answer)

File: kelly/스택 & 큐/Trapping Rain Water/Solution02.java
import java.util.*;

class Solution {
    public int trap(int[] height) {
        Deque<Integer> stack = new ArrayDeque<>();
        int[] water = new int[height.length];

        for (int i = 0; i < height.length; i++) {
            if (stack.isEmpty() || height[stack.peek()] >= height[i]) {
                stack.push(i);
                continue;
            }

            int lastPopped = -1;
            while (!stack.isEmpty() && height[stack.peek()] < height[i]) {
                lastPopped = stack.pop();
            }

            int lowWall = (stack.isEmpty()) ? lastPopped : i;
            int start = (stack.isEmpty()) ? lastPopped + 1 : stack.peek() + 1;

            for (int j = start; j < i; j++) {
                water[j] = height[lowWall] - height[j];
            }

            stack.push(i);
        }

        int answer = 0;
        for (int w : water) {
            answer += w;
        }

        return answer;
    }
}